File: backend/src/NLC.Application/DTOs/Jobs/JobDtos.cs
using NLC.Core.Enums;

namespace NLC.Application.DTOs.Jobs;

public record JobCardListDto(
    Guid     Id,
    string   JobNumber,
    string   WarehouseName,
    JobType  JobType,
    JobStatus Status,
    string   CustomerName,
    string?  ContainerNumber,
    string?  AsnNumber,
    string?  OrderNumber,
    string?  CurrentPhase,
    int      ProgressPercent,
    Priority Priority,
    bool     GrnGenerated,
    bool     ErpSynced,
    DateTime CreatedAt,
    DateTime? CompletedAt
);

public record JobCardDetailDto(
    Guid      Id,
    string    JobNumber,
    Guid      WarehouseId,
    string    WarehouseName,
    JobType   JobType,
    string    JobTypeConfigName,
    string[]  PhasesSnapshot,
    JobStatus Status,
    string    CustomerName,
    string?   ContainerNumber,
    string?   AsnNumber,
    string?   OrderNumber,
    string?   CurrentPhase,
    int       ProgressPercent,
    Priority  Priority,
    bool      GrnGenerated,
    bool      ErpSynced,
    string?   ReactivationReason,
    string?   ReactivatedBy,
    DateTime? ReactivatedAt,
    string?   Notes,
    DateTime  CreatedAt,
    DateTime? CompletedAt,
    List<PhaseLogDto>    PhaseLogs,
    List<ClockEventDto>  ClockEvents,
    List<SkuTallyDto>    SkuTallies,
    List<DispatchNoteDto> DispatchNotes
);

public record PhaseLogDto(
    Guid        Id,
    string      PhaseName,
    PhaseStatus PhaseStatus,
    bool        IsOptional,
    DateTime?   StartedAt,
    DateTime?   CompletedAt,
    string?     Notes
);

public record ClockEventDto(
    Guid      Id,
    Guid      WorkerId,
    string    WorkerName,
    string    PhaseName,
    DateTime  ClockInTime,
    DateTime? ClockOutTime,
    int?      DurationMinutes
);

public record SkuTallyDto(
    Guid         Id,
    string       SkuCode,
    string       SkuDescription,
    int          ExpectedQty,
    int          ScannedQty,
    int          TimeSpentMinutes,
    TallyStatus  TallyStatus,
    string       Source
);

public record DispatchNoteDto(
    Guid           Id,
    string         DnNumber,
    string         CustomerName,
    DispatchStatus DispatchStatus,
    DateTime       CreatedAt,
    DateTime?      DispatchedAt,
    List<DispatchSkuLineDto> SkuLines
);

public record DispatchSkuLineDto(
    Guid    Id,
    string  SkuCode,
    string  SkuDescription,
    int     OrderedQty,
    int     PickedQty,
    int     DispatchedQty,
    int     VarianceQty,
    bool    VarianceApproved,
    string? ApprovedBy
);

public record CreateJobCardRequest(
    Guid     WarehouseId,
    JobType  JobType,
    string   CustomerName,
    string?  ContainerNumber,
    string?  AsnNumber,
    string?  OrderNumber,
    Priority Priority,
    string?  Notes
);

public record CompletePhaseRequest(string PhaseName, string? Notes);
public record SkipPhaseRequest(string PhaseName, string Reason);
public record ReactivateJobRequest(string Reason, string ReactivatedBy);
public record ClockInRequest(Guid WorkerId, string PhaseName, string? Notes);
public record ClockOutRequest(Guid ClockEventId, string? Notes);

File: backend/src/main/java/com/nlc/web/dto/JobDtos.java
package com.nlc.web.dto;

import com.nlc.domain.enums.Enums.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

public final class JobDtos {
    private JobDtos() {}

    public record JobCardListDto(
        UUID id,
        String jobNumber,
        String warehouseName,
        String jobType,
        String status,
        String customerName,
        String containerNumber,
        String asnNumber,
        String orderNumber,
        String currentPhase,
        int progressPercent,
        String priority,
        boolean grnGenerated,
        boolean erpSynced,
        OffsetDateTime createdAt,
        OffsetDateTime completedAt
    ) {}

    public record JobCardDetailDto(
        UUID id,
        String jobNumber,
        UUID warehouseId,
        String warehouseName,
        String jobType,
        String jobTypeConfigName,
        String[] phasesSnapshot,
        String status,
        String customerName,
        String containerNumber,
        String asnNumber,
        String orderNumber,
        String currentPhase,
        int progressPercent,
        String priority,
        boolean grnGenerated,
        boolean erpSynced,
        String reactivationReason,
        String reactivatedBy,
        OffsetDateTime reactivatedAt,
        String notes,
        OffsetDateTime createdAt,
        OffsetDateTime completedAt,
        List<PhaseLogDto> phaseLogs,
        List<ClockEventDto> clockEvents,
        List<SkuTallyDto> skuTallies,
        List<DispatchNoteDto> dispatchNotes
    ) {}

    public record PhaseLogDto(
        UUID id,
        String phaseName,
        String phaseStatus,
        boolean isOptional,
        OffsetDateTime startedAt,
        OffsetDateTime completedAt,
        String notes
    ) {}

    public record ClockEventDto(
        UUID id,
        UUID workerId,
        String workerName,
        String phaseName,
        OffsetDateTime clockInTime,
        OffsetDateTime clockOutTime,
        Integer durationMinutes
    ) {}

    public record SkuTallyDto(
        UUID id,
        String skuCode,
        String skuDescription,
        int expectedQty,
        int scannedQty,
        int timeSpentMinutes,
        String tallyStatus,
        String source
    ) {}

    public record DispatchNoteDto(
        UUID id,
        String dnNumber,
        String customerName,
        String dispatchStatus,
        OffsetDateTime createdAt,
        OffsetDateTime dispatchedAt,
        List<DispatchSkuLineDto> skuLines
    ) {}

    public record DispatchSkuLineDto(
        UUID id,
        String skuCode,
        String skuDescription,
        int orderedQty,
        int pickedQty,
        int dispatchedQty,
        int varianceQty,
        boolean varianceApproved,
        String approvedBy
    ) {}

    // Requests
    public record CreateJobCardRequest(
        @NotNull UUID warehouseId,
        @NotBlank String jobType,
        @NotBlank String customerName,
        String containerNumber,
        String asnNumber,
        String orderNumber,
        String priority,
        String notes
    ) {}

    public record CompletePhaseRequest(@NotBlank String phaseName, String notes) {}
    public record SkipPhaseRequest(@NotBlank String phaseName, String reason) {}
    public record ReactivateJobRequest(@NotBlank String reason, @NotBlank String reactivatedBy) {}

    public record ClockInRequest(@NotNull UUID workerId, @NotBlank String phaseName, String notes) {}
    public record ClockOutRequest(@NotNull UUID clockEventId, String notes) {}
}
